central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,398 @@
+		indicatorDataStore: indicatorDataStore,
+	}
+}
+
+func (ds *datastoreImpl) AddProcessListeningOnPort(
+	ctx context.Context,
+	portProcesses ...*storage.ProcessListeningOnPortFromSensor,
+) error {
+	defer metrics.SetDatastoreFunctionDuration(
+		time.Now(),
+		"ProcessListeningOnPort",
+		"AddProcessListeningOnPort",
+	)
+
+	if !env.PostgresDatastoreEnabled.BooleanSetting() {
+		// PLOP is a Postgres-only feature, do nothing.
+		log.Warnf("Tried to add PLOP not on Postgres, ignore: %+v", portProcesses)
+		return nil
+	}
+
+	if ok, err := plopSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	portProcesses = normalizePLOPs(portProcesses)
+
+	// XXX: The next two calls, fetchIndicators and fetchExistingPLOPs, have to
+	// be done in a single join query fetching both ProcessIndicator and needed
+	// bits from PLOP.
+	indicatorsMap, indicatorIds, err := ds.fetchIndicators(ctx, portProcesses...)
+	if err != nil {
+		return err
+	}
+
+	existingPLOPMap, err := ds.fetchExistingPLOPs(ctx, indicatorIds, portProcesses...)
+	if err != nil {
+		return err
+	}
+
+	plopObjects := make([]*storage.ProcessListeningOnPortStorage, len(portProcesses))
+	for i, val := range portProcesses {
+		indicatorID := ""
+		var processInfo *storage.ProcessIndicatorUniqueKey
+
+		key := getProcesUniqueKey(val)
+
+		if indicator, ok := indicatorsMap[key]; ok {
+			indicatorID = indicator.GetId()
+			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
+		} else {
+			// XXX: Create a metric for this
+			log.Warnf("Found no matching indicators for %s", key)
+			processInfo = val.Process
+		}
+
+		plopKey := fmt.Sprintf("%d %d %s",
+			val.GetProtocol(),
+			val.GetPort(),
+			indicatorID,
+		)
+
+		existingPLOP, prevExists := existingPLOPMap[plopKey]
+
+		// There are three options:
+		// * We found an existing PLOP object with different close timestamp.
+		//   It has to be updated.
+		// * We found an existing PLOP object with the same close timestamp.
+		//   Nothing has to be changed (XXX: Ideally it has to be excluded from
+		//   the upsert later on).
+		// * No existing PLOP object, create a new one with whatever close
+		//   timestamp we have received and fetched indicator ID.
+		if prevExists && existingPLOP.CloseTimestamp != val.CloseTimestamp {
+			log.Debugf("Got existing PLOP: %+v", existingPLOP)
+
+			existingPLOP.CloseTimestamp = val.CloseTimestamp
+			existingPLOP.Closed = existingPLOP.CloseTimestamp != nil
+			plopObjects[i] = existingPLOP
+		}
+
+		if !prevExists {
+			if val.CloseTimestamp != nil {
+				// We try to close a not existing Endpoint, something is wrong
+				log.Warnf("Found no matching PLOP to close for %s", key)
+			}
+
+			newPLOP := &storage.ProcessListeningOnPortStorage{
+				// XXX, ResignatingFacepalm: Use regular GENERATE ALWAYS AS
+				// IDENTITY, which would require changes in store generator
+				Id:                 uuid.NewV4().String(),
```

Comment:
we've probably talked around the edges of this before, but what do you want the `Id` to be?  I understand not wanting to fix the generator now and I get not wanting to make it a custom store now.  But we need to consider what a future migration of the primary key would be if we were to change its type.  It is possible to likely that migration may be just as cumbersome to deal with as taking on the task of making the PK what you want it to be now.  Something to consider.

Reply:
My proposal was described in ROX-13793, to summarize the idea is to let the database to generate the id value. I know the migrations could be cumbersome in the future, but this one ideally should be easy, as it will be not a data migration (there still will be uuid), but only schema migration (with an extra logic how to generate uuid). Does it answer your concerns?